client/layout/masterbar/logged-in.jsx

```diff
@@ -400,6 +407,139 @@ class MasterbarLoggedIn extends Component {
+		const shouldShowPublicComingSoonSiteBadge =
+			! site.is_private &&
+			site.is_coming_soon &&
+			! isAtomicAndEditingToolkitDeactivated &&
+			! isTrial;
+
+		// Cover the coming Soon v1 cases for sites still unlaunched and/or in Coming Soon private by default.
+		// isPrivateAndUnlaunched means it is an unlaunched coming soon v1 site
+		const isPrivateAndUnlaunched = site.is_private && isUnlaunchedSite;
+		const shouldShowPrivateByDefaultComingSoonBadge = site.is_coming_soon || isPrivateAndUnlaunched;
+
+		// P2 Badge
+		if ( isSiteP2 && ! isP2Hub ) {
+			badges.push(
+				<span key="p2" className="masterbar__site-badge is-p2">
+					P2
+				</span>
+			);
+		}
+
+		// Staging Badge
+		if ( site?.is_wpcom_staging_site ) {
+			badges.push(
+				<span key="staging" className="masterbar__site-badge is-staging">
+					{ translate( 'Staging' ) }
+				</span>
+			);
+		}
+
+		// Trial Badge
+		if ( isTrial ) {
+			badges.push(
+				<span key="trial" className="masterbar__site-badge is-trial">
+					{ translate( 'Trial' ) }
+				</span>
+			);
+		}
+
+		// P2 Workspace Badge
+		if ( isP2Hub ) {
+			badges.push(
+				<span key="p2-workspace" className="masterbar__site-badge is-p2-workspace">
+					P2 Workspace
+				</span>
+			);
+		}
+
+		// Private/Coming Soon Badge
+		if ( site.is_private ) {
+			badges.push(
+				<span key="private" className="masterbar__site-badge is-private">
+					{ shouldShowPrivateByDefaultComingSoonBadge
+						? translate( 'Coming Soon' )
+						: translate( 'Private' ) }
+				</span>
+			);
+		}
+
+		// Express Service Badge
+		if ( site.options && site.options.is_difm_lite_in_progress ) {
+			badges.push(
+				<span key="express" className="masterbar__site-badge is-express">
+					{ translate( 'Express Service' ) }
+				</span>
+			);
+		}
+
+		// Public Coming Soon Badge
+		if ( shouldShowPublicComingSoonSiteBadge ) {
+			badges.push(
+				<span key="coming-soon" className="masterbar__site-badge is-coming-soon">
+					{ translate( 'Coming Soon' ) }
+				</span>
+			);
+		}
+
+		// Redirect Badge
+		if ( site.options && site.options.is_redirect ) {
+			badges.push(
+				<span key="redirect" className="masterbar__site-badge is-redirect">
+					{ translate( 'Redirect' ) }
+				</span>
+			);
+		}
+
+		// Domain Badge
+		if ( site.options && site.options.is_domain_only ) {
+			badges.push(
+				<span key="domain" className="masterbar__site-badge is-domain-only">
+					{ translate( 'Domain' ) }
+				</span>
+			);
+		}
+
+		// Site updates/errors indicators
+		if ( siteUpdates && ! siteIsConnected === false && siteUpdates.total > 0 ) {
+			badges.push(
+				<span key="updates" className="masterbar__site-badge is-updates">
+					{ translate( 'Updates' ) }
+				</span>
+			);
+		} else if ( siteIsConnected === false ) {
+			badges.push(
+				<span key="error" className="masterbar__site-badge is-error">
+					{ translate( 'Connection Error' ) }
+				</span>
+			);
+		}
+
+		return badges.length > 0 ? badges : null;
+	}
```

Comment:
It might be worth extracting these gadgets to a a `SiteBadges` component and reuse in both places.